### ISSUE TITLE
link error caused by `bsd_signal` if building with API level 21+ and uses libwebsocket

### ISSUE THREAD

**Issue by minggo** [cocos2d/cocos2d-x#16267]
If compiling with API level 21+, then it will cause link error that `bsd_signal` is missing if uses libwebsockets. It is because libwebsockets used `signal()`, and on Android 19 or lower version, `signal` is implemented as(can refer to `NDK_ROOT/platforms/android-19/arch-arm/usr/include/signal.h`)

``` c++
static __inline__ __sighandler_t signal(int s, __sighandler_t f)
{
    return bsd_signal(s,f);
}
```

But `bsd_signal` is removed with 21+. There is also [an issue](https://github.com/android-ndk/ndk/issues/160) in `android-ndk`. It seems it will be fixed in NDK r13.

**Comment by minggo** [comment]
But it is strange that, when i build with 19 and run on Nexus 5x(Android 6.0), it doesn't crash. But it may crash on some devices. I don't know why. I think it will crash on all devices with Android OS 5.0+(21).

**Comment by mastakrishna** [comment]
What's the crash log / call stack?

**Comment by mastakrishna** [comment]
If the bsd_signal is indeed missing in some devices libc.so, you may modify my code from http://stackoverflow.com/a/38660684 to dlsym "signal" if "bsd_signal" is missing. You may try using RTLD_NEXT/RTLD_DEFAULT or dlopen "libc.so" and dlsym from there.

**Comment by minggo** [comment]
@mastakrishna thanks, but i think can simplify it with 

``` c++
#if __ANDROID_API__ > 19
extern "C" {
  __sighandler_t bsd_signal(int s, __sighandler_t f) {

    return signal(s, f);
  }
}
#endif
```

Because 21+ has `signal()`.

**Comment by mastakrishna** [comment]
No, you can't. I tried it and it's not the way to go, because if you do so you'll loose compatibility with devices that don't have signal in their libc.so (API <= 19, as back then it was just an inline, so no symbol on the disk).
It'll compile and link fine, but the library will fail to load at runtime on older devices.

**Comment by minggo** [comment]
Yep, you are right. My method only can link but can not run under 21.

**Comment by mastakrishna** [comment]
FYI: I've updated StackOverflow code a bit to include #if condition for `__ANDROID_API__` and some comment about `dlsym`-ing `signal`if things go south.

**Comment by minggo** [comment]
fixed in #16292
